Refactoring build and test logic

pjbgf opened the pull request:
**Changes summary**
- Remove Dapper by relying on `docker buildx` and Makefile to download required tooling.
- Migrate away from Drone to GitHub actions.
- Refactor E2E testing logic. In CI it is now executed against versions `2.7`, `2.8` and `2.9`.
- External dependency versions (e.g. sonobuoy, kind, etc) are centralised at `hack/make/deps.mk`.

pjbgf committed:
build: Refactor project to remove Dapper
Moved Makefile targets logic from bash scripts within scripts/ to
the Makefile. The definition of versions for any required tooling
is now also centralised on Makefile.

The support for multi-architecture is based on docker buildx. But
for flexibility, both builder and runners can be defined so that
Podman support could be persued in the future.

Signed-off-by: Paulo Gomes <paulo.gomes@suse.com>

pjbgf committed:
build: Add GHA workflows

Signed-off-by: Paulo Gomes <paulo.gomes@suse.com>

pjbgf committed:
build: Refactor Makefile logic
The multi-stage dockerfile is now used to build the project binary, instead
of copying it from the host. This enforces that the binary is always rebuilt
whenever there is a change that invalidates the previous version - from a
container image perspective.

The makefile was restructured so that only the most important and relevant code
is present. In order to enable code reuse, build related logic was extracted to
hack/make/build.mk. All code related to installing tools is now placed in
hack/make/tools.mk.

Signed-off-by: Paulo Gomes <paulo.gomes@suse.com>

pjbgf committed:
build: Replace drone release with GHA

Signed-off-by: Paulo Gomes <paulo.gomes@suse.com>

pjbgf committed:
build: Remove persistent changes
Instead of changing the default docker buildx machine, create it and refer
to it by its name. This ensures building this project does not disrupt
other workflows from a given contributor.

Signed-off-by: Paulo Gomes <paulo.gomes@suse.com>

macedogm reviewed (commented):
Added some small nits and ideas.

Looks really good. Thanks for putting all the effort here.

pjbgf commented on package/Dockerfile:
```diff
@@ -1,39 +1,66 @@
 FROM registry.suse.com/bci/bci-micro:15.5 AS micro
 
+# Image that provides cross compilation tooling.
+FROM --platform=$BUILDPLATFORM tonistiigi/xx:latest as xx
```
This is created and maintained by a Docker employee, and the Docker ecosystem [relies on it](https://github.com/moby/buildkit/blob/b2796bb3517d602a5547d01eef99dcc957e181ea/Dockerfile#L40). But I agree that if we are to heavily rely on it, that should come from our mirror. Starting that process now.

pjbgf replied:
This has now been updated to use `rancher/mirrored-tonistiigi-xx:1.3.0` instead.

pjbgf commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,66 @@
+        go-version: 'stable'
+    - run: make validate
+
+  test:
```
That is a [pre-req](https://github.com/rancher/security-scan/blob/996449be80e32170da4c595b2e219802d3ffafba/Makefile#L57) for the `e2e` target, so it is running [later on](https://github.com/rancher/security-scan/blob/996449be80e32170da4c595b2e219802d3ffafba/.github/workflows/tests.yml#L66) this pipeline.

pjbgf committed:
build: Verify checksum for downloaded dependencies

Signed-off-by: Paulo Gomes <paulo.gomes@suse.com>

pjbgf commented on package/Dockerfile:
```diff
@@ -1,38 +1,77 @@
 RUN mkdir -p /chroot/etc/kube-bench/ && \
-    curl -sLf "https://github.com/aquasecurity/kube-bench/archive/refs/tags/v${kube_bench_version}.tar.gz" | \
-    tar xvz -C /chroot/etc/kube-bench/ --strip-components=1 "kube-bench-${kube_bench_version}/cfg"
+    curl -sLf "https://github.com/aquasecurity/kube-bench/archive/refs/tags/v${KUBE_BENCH_VERSION}.tar.gz" | \
```
Source code is immutable in GH releases, so unless the tag/release has been deleted and recreated, this should not change.

We should handle this separately - on a follow-up PR. An idea would be to download and sync upstream cfgs to our configs, so that we can go through the same validation processes we undergo for our own.

pjbgf replied:
cc @andypitcher wdyt?

andypitcher replied:
This is a good idea, this could be the beginning of automating the upgrade of our CIS profiles ? As soon as a new version of upstream CFGs is downloaded, new PRs are automatically created to make rke/k3s/rke2 (hardened/permissive) profiles. That might require `renovate`, maybe?

pjbgf replied:
We probably could do this independently of renovate. Maybe a new Make target that does that op, and a GHA workflow that creates a PR when it identifies changes on the worktree.

pjbgf commented on package/Dockerfile:
```diff
@@ -1,39 +1,66 @@
-    curl -sLf "https://github.com/aquasecurity/kube-bench/archive/refs/tags/v${kube_bench_version}.tar.gz" | \
-    tar xvz -C /chroot/etc/kube-bench/ --strip-components=1 "kube-bench-${kube_bench_version}/cfg"
+    curl -sLf "https://github.com/aquasecurity/kube-bench/archive/refs/tags/v${KUBE_BENCH_VERSION}.tar.gz" | \
+    tar xvz -C /chroot/etc/kube-bench/ --strip-components=1 "kube-bench-${KUBE_BENCH_VERSION}/cfg"
```
Great idea! Made changes to that effect. Points of note:
- Changed all dependencies and versions to their own file: `hack/make/deps.mk`, as `Makefile` started to look too cluttered. This will also help with not breaking renovate for the time being.
- Renovate supports this for github-releases. For things like `kubectl` we may need to find a better way around it (maybe `updatecli` or maybe a [custom datasource](https://docs.renovatebot.com/modules/datasource/custom/)).

pjbgf replied:
A new issue on renovate-config has been created to track the [checksums](https://github.com/rancher/renovate-config/issues/206) changes required.

pjbgf committed:
build: Source xx image from rancher/mirrored

Signed-off-by: Paulo Gomes <paulo.gomes@suse.com>

pjbgf committed:
build: Avoid version logic being executed twice
The version logic is contained within hack/make/build.mk and is a pre-req for some of the Makefile targets.

The Makefile is used to kick off the building of the container image, and inside that
process make build is executed to compile security-scan. This change stops the version
logic from executing a second time by accepting a VERSION var to be set, therefore removing
the need for accessing the .git dir.

Signed-off-by: Paulo Gomes <paulo.gomes@suse.com>

pjbgf commented:
Although the cross-compilation is [verified](https://github.com/rancher/security-scan/blob/4168e18654976560e27c5127ce88f189842917a7/package/Dockerfile#L77) as part of the new building process, the `arm64` binary/image was also manually tested:
```
$ uname -a
Linux rpi-top 6.1.21-v8+ #1642 SMP PREEMPT Mon Apr  3 17:24:16 BST 2023 aarch64 GNU/Linux

$ podman run --rm quay.io/paulinhu/security-scan:dev kb-summarizer --version
kb-summarizer version 4168e18-dirty
```
